express/dataset.py

```diff
@@ -0,0 +1,200 @@
+        index_columns = list(self.manifest.index.fields.keys())
+        # load index dataframe
+        index_df = df[index_columns]
+
+        self._upload_index(index_df)
+
+    def add_subsets(self, df: dd.DataFrame, spec: ComponentSpec):
+        for name, subset in spec.output_subsets.items():
+            fields = list(subset.fields.keys())
+            # verify fields are present in the output dataframe
+            subset_columns = [f"{name}_{field}" for field in fields]
+            for col in subset_columns:
+                if col not in df.columns:
+                    raise ValueError(
+                        f"Column {col} present in output subsets but not found in dataset"
+                    )
+
+            # load subset dataframe
+            subset_df = df[subset_columns]
+            # add to the manifest and upload
+            self._upload_subset(name, subset.fields, subset_df)
+
+    def upload(self, save_path):
+        Path(save_path).parent.mkdir(parents=True, exist_ok=True)
+        self.manifest.to_file(save_path)
+        return None
+
+
+class FondantComponent:
+    def __init__(self, type="transform"):
+        # note: Fondant spec always needs to be called like this
+        # and placed in the src directory
+        self.spec = ComponentSpec("fondant_component.yaml")
+        self.type = type
+
+    def run(self) -> dd.DataFrame:
+        """
+        Parses input data, executes the transform, and creates the output manifest.
+
+        Returns:
+            Manifest: the output manifest
+        """
+        # step 1: add and parse arguments
+        args = self._add_and_parse_args()
+        # step 2: create Fondant dataset based on input manifest
+        metadata = json.loads(args.metadata)
+        if self.type == "load":
+            # TODO ideally get rid of args.metadata
+            # by including them in the storage args,
+            # getting run_id based on args.output_manifest_path
+            manifest = Manifest.create(
+                project_name=metadata["project_name"],
+                base_path=metadata["base_path"],
+                run_id=metadata["run_id"],
+                component_id=metadata["component_id"],
+            )
+        else:
+            manifest = Manifest.from_file(args.input_manifest_path)
+        dataset = FondantDataset(manifest)
+        # step 3: load or transform data
+        if self.type == "load":
+            df = self.load(args)
+            dataset.add_index(df)
+            dataset.add_subsets(df, self.spec)
+        else:
+            # create HF dataset, based on component spec
+            input_dataset = dataset.load_data(self.spec)
+            # provide this dataset to the user
+            df = self.transform(
+                dataset=input_dataset,
+                args=args,
+            )
+
+        # step 4: create output manifest
+        output_manifest = dataset.upload(save_path=args.output_manifest_path)
```

Comment:
This method doesn't return anything.

But in any case, I think we should evolve the manifest from the dataset separately. We should be able to deduct the output manifest from just the input manifest and the component spec, which we can then use for validation of the output dataset. This functionality will also be useful for static analysis in the future.

We can rework this in a separate PR.